### PR TITLE
feat(rubygems): fallback to info when version fails

### LINKED ISSUE #7320: Artifactory doesn't support versions list for Rubygems
**What Renovate type, platform and version are you using?**

self hosted 23.31.1

**Describe the bug**

After we fixed that the wrong url was used for rubygems in artifactory we are now facing the problem that renovate tries to do a request to `https://bahnhub.tech.rz.db.de/artifactory/api/gems/seec-mov-gems-prod-local/api/v1/versions/sfn_seec_base.json`

this version endpoint doesnt seem to be supported by artifactory see:
https://www.jfrog.com/jira/browse/RTFACT-16005
https://www.jfrog.com/jira/browse/RTFACT-19012

as i'm not a ruby developer i'm not sure how important these are but for me they sound like a showstopper. 

**Relevant debug logs**
```
DEBUG: Datasource unknown error (repository=seec-mop/mopro/move/templates)
       "datasource": "rubygems",
       "lookupName": "sfn_seec_base",
       "err": {
         "name": "HTTPError",
         "timings": {
           "start": 1600427716942,
           "socket": 1600427716942,
           "lookup": 1600427716942,
           "connect": 1600427716988,
           "secureConnect": 1600427717007,
           "upload": 1600427717007,
           "response": 1600427717027,
           "end": 1600427717028,
           "phases": {
             "wait": 0,
             "dns": 0,
             "tcp": 46,
             "tls": 19,
             "request": 0,
             "firstByte": 20,
             "download": 1,
             "total": 86
           }
         },
         "message": "Response code 400 (Bad Request)",
         "stack": "HTTPError: Response code 400 (Bad Request)\n    at Request.<anonymous> (/Users/henrysachs/.nvm/versions/node/v14.11.0/lib/node_modules/renovate/node_modules/got/dist/source/as-promise/index.js:117:42)\n    at processTicksAndRejections (internal/process/task_queues.js:93:5)",
         "options": {
           "headers": {
             "user-agent": "https://github.com/renovatebot/renovate",
             "accept-encoding": "gzip, deflate",
             "hosttype": "rubygems",
             "authorization": "***********",
             "accept": "application/json"
           },
           "url": "https://bahnhub.tech.rz.db.de/artifactory/api/gems/seec-mov-gems-prod-local/api/v1/versions/sfn_seec_base.json",
           "username": "",
           "password": "***********",
           "method": "GET",
           "http2": false
         },
         "response": {
           "statusCode": 400,
           "statusMessage": "Bad Request",
           "body": {"errors": [{"status": 400, "message": "Bad Request"}]},
           "headers": {
             "date": "Fri, 18 Sep 2020 11:15:17 GMT",
             "content-type": "application/json",
             "transfer-encoding": "chunked",
             "connection": "close",
             "server": "Artifactory/6.11.7",
             "x-artifactory-id": "98525d72110e9d1d58e52368ccbf330d505d881f",
             "x-artifactory-node-id": "i-0e67682b644269f1c",
             "x-frame-options": "DENY",
             "x-content-type-options": "nosniff",
             "x-xss-protection": "1; mode=block"
           }
         }
       }
```

**To Reproduce**

setup an artifactory and try to pull your dependencies from there. see also #7315 

**Additional context**

none

### PR DESCRIPTION
**Pull request by henrysachs**
when the version endpoint fails use the single version from the info endpoint.
This occurs for example when using artifactory.

Closes #7320

### CODE REVIEW

**Comment by henrysachs** [comment]
not quite sure why macos fails on install when i didnt even changed the package.json

**Comment by henrysachs** [comment]
and im not quite sure if i need to wrap the await into a try catch because i only worked with promises and observables until now.

**Comment by henrysachs** [comment]
@rarkins not quite sure how to tag you in this here i'm more of a gitlab guy. Please don't feel annoyed by this would just be happy to get feedback.

**Comment by henrysachs** [comment]
just found #7111  @micheelengronne  maybe we could somehow combine our efforts in this area?

**Comment by henrysachs** [comment]
but the problem in my case seems to be that artifactory doesnt emit the dependencies as json. Instead it seems to be some sort of file that i dont know how to open to properly read it. see last request on here this site: 
https://guides.rubygems.org/rubygems-org-api/#gem-version-methods

**Comment by rarkins** [comment]
So Artifactory does have an endpoint that lists all available versions?

**Comment by henrysachs** [comment]
i just found out that you can misuse the dependencies endpoint for this 😅 but the problem here is that it is in this marshalled array format and its a file you need to download. not JSON or XML. Just found it because I found the MR mentioned above.

**Comment by micheelengronne** [comment]
Thanks @henrysachs I dont' currently have time to finish my MR. It was intended to be used with the Gemstash repository. Feel free to take what you need from it.

**Comment by rarkins** [comment]
I think there's value in falling back to the info endpoint "latest" result as-is and then adding additional versions retrieval after.

**Comment by henrysachs** [comment]
of course. Maybe i can do a seperate pr for that. But i currently dont have an idea how to properly parse this ruby marshalled array.

**Comment by henrysachs** [comment]
So just to come back to this one what do you think about it?

File: lib/datasource/rubygems/get.ts [hunk 1]
```diff
@@ -52,7 +52,7 @@ export async function getDependency(
 
   const versions = (await fetch(dependency, registry, VERSIONS_PATH)) || [];
```
Comment (rarkins) [inline lib/datasource/rubygems/get.ts:53]: Doesn't this request *throw* ? If so then everything below it is not executed.

Comment (henrysachs) [reply lib/datasource/rubygems/get.ts:53]: i thought it doesnt because of the ||  if not than i can add a try catch around it

Comment (henrysachs) [reply lib/datasource/rubygems/get.ts:53]: just to be sure a try catch would look like the following right? I can implement this if you want me to.

```js
let versions = []
try {
 versions = await fetch(dependency, registry, VERSIONS_PATH);
} catch {
// all of my code
}
```

**Comment by henrysachs** [comment]
so i just catched the error and moved some code to match the flow better. Now i got everything working and received a pull request for this repo: https://github.com/henrysachs/ruby-test-1337/pull/8 
🎉 

also found out that artifactory is just pulling "through" to the version endpoint if its just a mirrored dependency so i had to manually tweak the version path that it erros on purpose.

**Comment by henrysachs** [comment]
when everything is fine i would like to rebase to squash all commits into 1 and add myself to the contributors 😅

File: lib/datasource/rubygems/get.ts [hunk 2]
```diff
@@ -50,23 +50,13 @@ export async function getDependency(
...
-      releaseTimestamp,
-      rubygemsVersion,
-      rubyVersion,
-    })
-  );
+  let versions = [];
+  let releases = [];
+  try {
+    versions = await fetch(dependency, registry, VERSIONS_PATH);
+  } catch (err) {
```
Comment (rarkins) [inline lib/datasource/rubygems/get.ts:57]: We need to check what type of error it is, and only continue if it's the one we expect from Artifactory, e.g. 404. If instead we get a 401 or 5xx response or connection failed, we *don't* want to continue and instead throw an error like usual.

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/datasource/rubygems/get.ts (hunk 2 above)
Comment (henrysachs) [reply lib/datasource/rubygems/get.ts:57]: done

File: lib/datasource/rubygems/get.ts [hunk 3]
```diff
@@ -50,23 +50,47 @@ export async function getDependency(
...
 
-  const versions = (await fetch(dependency, registry, VERSIONS_PATH)) || [];
+  let versions = [];
+  let releases = [];
+  try {
+    versions = await fetch(dependency, registry, VERSIONS_PATH);
+  } catch (err) {
+    if (err.statusCode === 400 || err.statusCode === 404) {
+      logger.debug({ registry }, 'version endpoint errors or is not available');
+      logger.debug({ err });
```
Comment (rarkins) [inline lib/datasource/rubygems/get.ts:60]:
```suggestion
      logger.debug({ registry }, 'versions endpoint returns error - falling back to info endpoint');
```

File: lib/datasource/rubygems/get.ts [hunk 4]
```diff
@@ -50,23 +50,47 @@ export async function getDependency(
...
-  );
+  if (versions.length === 0 && info.version) {
+    logger.warn('falling back to the version from the info endpoint');
+    releases = [
+      {
+        version: info.version,
+        rubyPlatform: info.platform,
+        releaseTimestamp: null,
+        rubyVersion: null,
+        rubygemsVersion: '\u003e= 0',
```
Comment (rarkins) [inline lib/datasource/rubygems/get.ts:74]: What is the meaning of this value?

Comment (henrysachs) [reply lib/datasource/rubygems/get.ts:74]: it was the default value from the versions endpoint its basically >= 0

Comment (henrysachs) [reply lib/datasource/rubygems/get.ts:73]: but we can remove it I just wanted to match the signature from the "default" case

Comment (rarkins) [reply lib/datasource/rubygems/get.ts:73]: Yes, remove it

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs tests to get coverage back to 100%, also run `yarn lint-fix`

**Comment by henrysachs** [comment]
never done something like that is there somewhere you can point me to?

**Comment by rarkins** [comment]
> never done something like that is there somewhere you can point me to?

Test files are colocated with source files and you can find some existing rubygems tests already that you can duplicate then modify

**Comment by henrysachs** [comment]
added tests and run yarn lint fix.

**Review by rarkins — APPROVED** [review]

**Comment by henrysachs** [comment]
What happens after the approval? 😅

**Comment by rarkins** [comment]
It needs to be up to date and pass all tests before we merge

**Comment by henrysachs** [comment]
Thanks! So happy for your help to get me through this. 🎉

**Pull request merged**